### PR TITLE
supplet several interface of static Variable to consistent with dygraph Tensor

### PR DESCRIPTION
**Pull request by CtfGo**
### PR types

New features

### PR changes

APIs

### Describe

supplet several interface of static Variable to consistent with dygraph Tensor，including 1 property(ndim) and 4 methods( ndimension,dim,size,__matmul__,detach)， the preview of api document listed below：
image
image
image
image

### CODE REVIEW

**Review by Aurelius84 — DISMISSED** [review]
LGTM

**Review by Aurelius84 — DISMISSED** [review]

File: paddle/fluid/operators/share_data_op.cu [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zhhsplendid) [inline paddle/fluid/operators/share_data_op.cu]: You can change in next PR:

2019 -> 2021

File: python/paddle/fluid/framework.py [hunk 2]
```diff
@@ -947,35 +947,45 @@ def __init__(self,
...
         self.is_data = is_data
 
-    @fake_interface_only
     def detach(self):
         """
-        **Notes**:
-            **This API is ONLY available in Dygraph mode**
-
         Returns a new Variable, detached from the current graph.
+        It will share data with origin Variable and always doesn't have a Tensor copy.
```
Comment (zhhsplendid) [inline python/paddle/fluid/framework.py]: "and always doesn't have a Tensor copy" -> "without tensor copy"

File: python/paddle/fluid/tests/unittests/test_math_op_patch.py [hunk 3]
```diff
@@ -336,6 +335,28 @@ def test_bitwise_not(self):
...
+    @prog_scope()
+    def test_ndim(self):
+        a = paddle.static.data(name="a", shape=[10, 1])
+        self.assertEqual(a.dim(), 2)
+        self.assertEqual(a.ndimension(), 2)
+        self.assertEqual(a.ndim, 2)
+
+    @prog_scope()
+    def test_matmul(self):
+        a = fluid.layers.data(name='a', shape=[2, 3], dtype='float32')
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_math_op_patch.py]: You can change in next PR:

use 2.0 api instead of fluid API for the new code.

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 4]
```diff
@@ -345,6 +344,52 @@ def _test():
 
         self.assertRaises(Exception, _test)
 
+    def test_size(self):
+        prog = paddle.static.Program()
+        with fluid.program_guard(prog):
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_variable.py]: You can change in next PR:

use 2.0 api instead of fluid API for the new code.

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 5]
```diff
@@ -345,6 +344,52 @@ def _test():
...
+
+        startup = fluid.Program()
+        main = paddle.static.Program()
+        scope = fluid.core.Scope()
+        with fluid.scope_guard(scope):
+            with fluid.program_guard(main, startup):
+                x = fluid.data(name='x', shape=[3, 2, 1], dtype='float32')
+                x.persistable = True
+                feed_data = np.ones(shape=[3, 2, 1], dtype=np.float32)
+                detach_x = x.detach()
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_variable.py:379]: Could we also test:
1. Change to `detach_x` would also change `x`
2. Change to `x` would also change `detach_x`

**Review by zhhsplendid — DISMISSED** [review]
I gave some comments, but you can change in next PR since you already passed some CIs.

I will approve this PR.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 5 above)
Comment (CtfGo) [reply python/paddle/fluid/tests/unittests/test_variable.py:379]: I have updated code with these tips, thank you.

**Review by zhhsplendid — DISMISSED** [review]
LGTM

File: python/paddle/fluid/framework.py [hunk 6]
```diff
@@ -1810,6 +1820,35 @@ def set_value(self, value, scope=None):
...
+    def size(self):
+        """
+        Returns the number of elements for current Variable, which is a int64 Variable with shape [1]
+
+        Returns:
+            Variable: the number of elements for current Variable
+
+        Examples:
+            .. code-block:: python
+            :name: code-example1
```
Comment (chenwhql) [inline python/paddle/fluid/framework.py]: 有文档预览的结果吗？麻烦将结果截图放到PR里

File: paddle/fluid/operators/share_data_op.cc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            "The type of input (Input) and output (Out) are inconsistent."));
+
+    ctx->ShareDim("Input", "Out");
+  }
+};
+
+class ShareDataOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Input", "The input tensor.");
```
Comment (chenwhql) [inline paddle/fluid/operators/share_data_op.cc]: The input tensor of ShareData operator？

File: python/paddle/fluid/framework.py (hunk 6 above)
Comment (CtfGo) [reply python/paddle/fluid/framework.py]: done

File: paddle/fluid/operators/share_data_op.cc (hunk 7 above)
Comment (CtfGo) [reply paddle/fluid/operators/share_data_op.cc]: yes, used on static.Variable.detach

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM
